src/k8s/pkg/k8sd/features/cilium/gateway_test.go

```diff
@@ -0,0 +1,278 @@
+package cilium_test
+
+import (
+	"context"
+	"errors"
+	"testing"
+
+	"github.com/canonical/k8s/pkg/client/helm"
+	helmmock "github.com/canonical/k8s/pkg/client/helm/mock"
+	"github.com/canonical/k8s/pkg/client/kubernetes"
+	"github.com/canonical/k8s/pkg/k8sd/features/cilium"
+	"github.com/canonical/k8s/pkg/k8sd/types"
+	snapmock "github.com/canonical/k8s/pkg/snap/mock"
+
+	. "github.com/onsi/gomega"
+	v1 "k8s.io/api/apps/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/kubernetes/fake"
+	"k8s.io/utils/ptr"
+)
+
+func TestGatewayEnabled(t *testing.T) {
+	t.Run("HelmApplyErr", func(t *testing.T) {
+		g := NewWithT(t)
+
+		applyErr := errors.New("failed to apply")
+		helmM := &helmmock.Mock{
+			ApplyErr: applyErr,
+		}
+		snapM := &snapmock.Snap{
+			Mock: snapmock.Mock{
+				HelmClient: helmM,
+			},
+		}
+		network := types.Network{}
+		gateway := types.Gateway{
+			Enabled: ptr.To(true),
+		}
+
+		status, err := cilium.ApplyGateway(context.Background(), snapM, gateway, network, nil)
+
+		g.Expect(err).To(HaveOccurred())
+		g.Expect(err.Error()).To(ContainSubstring(applyErr.Error()))
+		g.Expect(status.Enabled).To(BeFalse())
+		g.Expect(status.Version).To(Equal(cilium.CiliumAgentImageTag))
+		g.Expect(status.Message).To(ContainSubstring("Failed to deploy Cilium Gateway"))
+		g.Expect(helmM.ApplyCalledWith).To(HaveLen(1))
+
+	})
+
+	t.Run("AlreadyDeployed", func(t *testing.T) {
+		g := NewWithT(t)
+
+		helmM := &helmmock.Mock{
+			ApplyChanged: false,
+		}
+		snapM := &snapmock.Snap{
+			Mock: snapmock.Mock{
+				HelmClient: helmM,
+			},
+		}
+		network := types.Network{}
+		gateway := types.Gateway{
+			Enabled: ptr.To(true),
+		}
+
+		status, err := cilium.ApplyGateway(context.Background(), snapM, gateway, network, nil)
+
+		g.Expect(err).NotTo(HaveOccurred())
+		g.Expect(status.Enabled).To(BeTrue())
+		g.Expect(status.Version).To(Equal(cilium.CiliumAgentImageTag))
+		g.Expect(status.Message).To(Equal(cilium.EnabledMsg))
+
+		helmCiliumArgs := helmM.ApplyCalledWith[2]
+		g.Expect(helmCiliumArgs.Chart).To(Equal(cilium.ChartCilium))
+		g.Expect(helmCiliumArgs.State).To(Equal(helm.StateUpgradeOnly))
+		g.Expect(helmCiliumArgs.Values["gatewayAPI"].(map[string]any)["enabled"]).To(Equal(true))
+
+	})
+
+	t.Run("RolloutFail", func(t *testing.T) {
+		g := NewWithT(t)
+
+		helmM := &helmmock.Mock{
+			ApplyChanged: true,
+		}
+		clientset := fake.NewSimpleClientset()
+		snapM := &snapmock.Snap{
+			Mock: snapmock.Mock{
+				HelmClient: helmM,
+				KubernetesClient: &kubernetes.Client{
+					Interface: clientset,
+				},
+			},
+		}
+		network := types.Network{}
+		gateway := types.Gateway{
+			Enabled: ptr.To(true),
+		}
+
+		status, err := cilium.ApplyGateway(context.Background(), snapM, gateway, network, nil)
+
+		g.Expect(err).To(HaveOccurred())
+		g.Expect(err.Error()).To(ContainSubstring("rollout restart cilium to enable Gateway API"))
```

Comment:
I'm personally not a fan of checking for error strings, unless we're making sure we see some error that we passed in that test specific test case. Like `applyErr`s in our feature tests.